static/js/message_store.js

```diff
@@ -128,19 +129,21 @@ function add_message_metadata(message) {
 
     // Add new people involved in this message to the people list
     _.each(involved_people, function (person) {
-        // Do the hasOwnProperty() call via the prototype to avoid problems
-        // with keys like "hasOwnProperty"
-        if (! people.get_by_email(person.email)) {
-            people.add(person);
-        }
-
-        if (people.get_by_email(person.email).full_name !== person.full_name) {
-            people.reify(person);
-        }
+        if (!person.unknown_local_echo_user) {
+            if (! people.get_by_email(person.email)) {
+                people.add({
+                    email: person.email,
+                    user_id: person.user_id || person.id,
+                    full_name: person.full_name,
+                    is_admin: person.is_realm_admin || false,
+                    is_bot: person.is_bot || false
```

Comment:
I don't like it either, but the ugliness precedes my change; it's just more overt now.  The problem here is that we are trying to update our list of people with data structures that aren't really intended for people-updates.  There are a few ways to fix this. One is that we could have the back end send is_bot, is_realm_admin, and user_id on all messages.  But I think a cleaner approach would be to send add-people events as a separate event.  The tricky part about the latter approach, unfortunately, is that I don't think the back end necessarily knows when it's sending an "unknown" person to the browser.  Finally, we could just have the browser fetch people explicitly when it gets an email that it hasn't encountered.